fix(server): Cache glob patterns to improve RBAC evaluation performance

Sinhyeok opened the pull request:
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates? **Yes**
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

---

Fixes #25757

## Summary

Cache compiled glob patterns in `util/glob/glob.go` using a bounded LRU cache to eliminate redundant `glob.Compile()` calls during RBAC policy evaluation.

## Changes

- Add LRU cache (`golang/groupcache/lru`) for compiled glob patterns with bounded size
- Default cache size: 10,000 patterns (configurable via `ARGOCD_GLOB_CACHE_SIZE`)
- Add explicit mutex for thread-safety
- Add unit tests for caching behavior, concurrent access, and LRU eviction
- Add benchmarks demonstrating performance improvement
- Add documentation in `high_availability.md` and `rbac.md`

## Performance Improvement

| Metric | Before | After | Improvement |
|--------|--------|-------|-------------|
| **Latency** | 1,917 ns/op | 39 ns/op | **~49x faster** |
| **Memory** | 1,270 B/op | 0 B/op | **100% reduction** |
| **Allocations** | 34 allocs/op | 0 allocs/op | **100% reduction** |

## Benchmark Results

```
BenchmarkMatch_WithCache-14         	29898592	    39.27 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatch_WithoutCache-14      	  657784	  1917.00 ns/op	    1270 B/op	      34 allocs/op
BenchmarkGlobCompile-14             	  907964	  1385.00 ns/op	    1240 B/op	      33 allocs/op
BenchmarkMatch_RBACSimulation-14    	 1000000	  1045.00 ns/op	       0 B/op	       0 allocs/op
```

## Why LRU Cache?

- **Bounded size**: Prevents memory exhaustion from untrusted RBAC patterns
- **LRU eviction**: Frequently used patterns stay cached, rarely used ones are evicted
- **Configurable**: `ARGOCD_GLOB_CACHE_SIZE` environment variable for tuning

Sinhyeok committed:
fix: memoize compiled glob patterns to improve RBAC performance

Glob patterns in util/glob/glob.go were being compiled on every Match()
call without caching. In environments with large RBAC policies (~6,900
policy lines) and many applications (~3,000), this caused
ApplicationService/List API calls to take 30-40 seconds.

Root cause: Casbin policy evaluation calls glob.Match() for each
policy rule check. With the policy effect rule requiring deny checks,
all policies must be scanned, resulting in ~150,000 glob.Compile()
calls per List request.

Solution: Use sync.Map to memoize compiled glob patterns. sync.Map is
optimized for the 'write once, read many' pattern which matches our
use case perfectly.

Performance improvement:
- glob.Compile() calls: 150,000 → ~1,940 (unique patterns only)
- ApplicationService/List latency: 30-40s → <1s
- Memory allocations per match: eliminated (0 B/op, 0 allocs/op)

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
fix(server): cache compiled glob patterns in RBAC evaluation

Glob patterns used in RBAC policy evaluation were being recompiled
on every match operation. For deployments with many applications and
complex RBAC policies, this caused significant latency (30+ seconds)
when the RBAC ConfigMap cache was invalidated.

This change caches compiled glob patterns using an LRU cache with
bounded size (default: 10,000 patterns, configurable via
ARGOCD_GLOB_CACHE_SIZE). This prevents memory exhaustion from
untrusted patterns while providing significant performance benefits.

Performance improvement:
- Before: ~2.8µs per match (with compilation)
- After: ~52ns per match (cache hit)
- ~50x speedup for cached patterns

Fixes #25757

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
docs: update comments from memoization to caching

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

nitishfy reviewed (changes requested):
Left a few comments, PTAL.

Sinhyeok commented on util/glob/glob_test.go:
```diff
@@ -86,3 +113,122 @@ func Test_MatchWithError(t *testing.T) {
+		go func() {
+			defer wg.Done()
+			result := Match(pattern, text)
+			require.True(t, result)
```
Good catch! Updated to use a channel for collecting errors from goroutines instead of calling require.True() directly.

Sinhyeok commented on util/glob/glob_test.go:
```diff
@@ -86,3 +113,122 @@ func Test_MatchWithError(t *testing.T) {
+	require.Equal(t, 1, getCacheLen(), "should only have one cached entry for the pattern")
+}
+
+func Test_GlobCacheLRUEviction(t *testing.T) {
```
Yes, added Test_InvalidGlobNotCached to verify that invalid patterns are not cached after failed compilation.

Sinhyeok committed:
test: address review feedback for glob cache tests

- Use channel for error collection in concurrent test (testing.T is not goroutine-safe)
- Add assertion for oldest entry eviction in LRU test
- Add test for invalid glob pattern not being cached
- Rename helper functions for clarity

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
Merge branch 'master' into fix/memoize-glob-patterns

ppapapetrou76 reviewed (approved):
LGTM

nitishfy reviewed (commented):
Left a few comments otherwise LGTM!

Sinhyeok committed:
test(util/glob): Assert single compile in concurrent cache test

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok commented on util/glob/glob_test.go:
```diff
@@ -86,3 +114,157 @@ func Test_MatchWithError(t *testing.T) {
+	require.Equal(t, result1, result2)
+}
+
+func Test_GlobCachingConcurrent(t *testing.T) {
```
Thanks. I added a compile counter by wrapping the compile function and now assert it only compiles once under concurrent access.

Sinhyeok committed:
test(util/glob): Assert cache hit in glob tests

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok commented:
> Left a few comments otherwise LGTM!

Thanks! I addressed the comments — added explicit cache-hit assertions (compile counted once) in both the sequential and concurrent tests.

Sinhyeok committed:
Merge branch 'master' into fix/memoize-glob-patterns

Signed-off-by: Sinhyeok Seo <44961659+Sinhyeok@users.noreply.github.com>

Sinhyeok committed:
chore: modernize glob_test.go to resolve golangci-lint errors

Use range-over-int syntax and WaitGroup.Go to satisfy rangeint and
waitgroup modernize linter rules.

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok commented:
> @Sinhyeok
> 
> LGTM, but it looks like there’s a merge conflict again.
> 
> Could you please resolve it? I’ll mark it as `ready-for-review` once it’s fixed.

@choejwoo 

Done! I’ve resolved the merge conflicts. Please take another look when you have a chance. Thanks!

Sinhyeok committed:
fix(glob): include separators in cache key to prevent incorrect matches

The glob cache was using only the pattern string as the key, but the
same pattern compiled with different separators produces different globs.
This could cause incorrect match results when the same pattern is used
with and without separators.

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
Merge branch 'master' into fix/memoize-glob-patterns

Sinhyeok committed:
refactor(util/glob): Simplify cache key construction

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
Merge branch 'master' into fix/memoize-glob-patterns

Sinhyeok committed:
refactor(util/glob): use struct for cache key to avoid collision risk

Use a globCacheKey struct instead of string concatenation with null
character separator. This eliminates the theoretical possibility of
key collisions since ConfigMaps can technically contain null characters.

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
refactor(util/glob): pipe glob cache size through argocd-cmd-params-cm

Instead of reading ARGOCD_GLOB_CACHE_SIZE directly from env in init(),
configure glob cache size via argocd-cmd-params-cm ConfigMap key
(server.glob.cache.size) and the --glob-cache-size server flag,
consistent with how other server config values are piped.

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
refactor(util/glob): use RWMutex and compile outside lock to reduce contention

Use sync.RWMutex instead of sync.Mutex so concurrent cache hits can
proceed in parallel. Move the expensive glob compilation outside the
lock using a double-check pattern, preventing unrelated patterns from
serializing against each other during warm-up or RBAC policy reloads.

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
fix(util/glob): bypass cache in BenchmarkMatch_WithoutCache

Call glob.Compile + Match directly instead of going through the
caching layer, so the benchmark measures pure compile-per-call cost
without lock acquisition, cache insertion, or LRU eviction overhead.

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
fix: use MaxInt32 for glob-cache-size upper bound

math.MaxInt is architecture-dependent and excessively large for a cache
size parameter. Use math.MaxInt32 for consistency with other flags in
the codebase and to provide a safer upper bound.

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
chore: regenerate docs and manifests for glob-cache-size flag

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
fix(util/glob): use singleflight to prevent duplicate glob compilations

Replace the RWMutex double-check pattern with singleflight to deduplicate
concurrent compilations of the same glob pattern. This fixes the flaky
Test_GlobCachingConcurrent test where multiple goroutines could compile
the same pattern before any cached it.

Also switch from RWMutex to Mutex because lru.Cache.Get() promotes entries
(MoveToFront), making it unsafe under RLock.

Signed-off-by: Sinhyeok Seo <sinhyeok@gmail.com>

Sinhyeok committed:
Merge branch 'master' into fix/memoize-glob-patterns

# Conflicts:
#	go.mod
